### ISSUE TITLE
Unpacking of iterables with unknown/wrong length

### ISSUE THREAD

**Issue by not-my-profile** [microsoft/pyright#8290]
It would be nice if Pyright could have a diagnostic rule to detect e.g. the following:

```python
def foo(x: list[int]):
    a, b = x
```

which will fail with a `ValueError` if there aren't enough or too many values to unpack.

This would encourage developers to use tuple types in parameters when a specific length is required.

Pyright currently also doesn't detect wrong unpacking of literals e.g:

```python
a, b = "abc"
```

Mypy detects this by forbidding string unpacking in general (`Unpacking a string is disallowed`), which does however mean mypy flags `a, b = "ab"` as a false positive.

Especially when iterating over iterables of union types this type checking would be helpful, e.g Pyright currently doesn't flag the problem in:

```python
def foo(x: list[tuple[int, int] | Literal["foo"]):
    for a, b in x:
        pass
```
